### DISCUSSION TITLE
Language Server Protocol

### DISCUSSION THREAD

**Discussion by mike-vogel** [textX/textX#339] [Q&A]
I need LSP support for my DSL. The initial implementation is in PyParser. Upgrading to textX looks like a great step forward for my DSL. It appears that the textX LS project is paused. Are there any other approaches for LSP support possible for textX based language?

**Comment by igordejanovic** [comment]
Yeah, textX-LS is in a dormant state at the moment. Until @danixeee  or someone else find some time to work on textX-LS your best bet is to use [pygls](https://github.com/openlawlibrary/pygls) which is a generic Python LSP and implement LSP for you DSL. It's a bit more work but on the other side you have full control over the LSP implementation. You can still use textX-LS as a reference as pygls is used there too.

**Comment by mike-vogel** [comment]
Thanks for the pointer to pygls and the guidance that looking into textX-LS is a way to start. Perhaps we can add to textX-LS after doing that research rather than creating a dedicated LSP server for our DSL.

**Comment by igordejanovic** [comment]
@mike-vogel That would be great!

**Comment by mike-vogel** [comment]
Other than using pygls and a LSP approach are there other examples or recommendations for building an editing environment for either the web (preferred) or desktop for a textX based language? I haven't had time to review how textX-LS does tab-completion and other editing help but building helpful completion seems hard. I started down the MPS path but the desktop only limitations of the current production versions have caused me to switch back to an LSP approach.

**Comment by igordejanovic** [comment]
Unfortunately, besides textX-LS which is still unfinished we don't have yet any out-of-the-box solution for building editing environments for textX languages. If someone has made any progress along these lines I would be very much interested to hear about it.

If you decide to go MPS route there has been some work to bring MPS to the web by @ftomassetti. Check out https://github.com/Strumenta/webeditkit
I haven't try it but maybe Federico can say what is the current state of the project.

**Comment by ftomassetti** [comment]
hi @igordejanovic thank you for including me in the conversation. I have discussed with @mike-vogel about WebEditKit, but at this stage the project is not mature enough for production
